cfme/tests/intelligence/reports/test_validate_chargeback_report.py

```diff
@@ -54,667 +46,308 @@
-
-    yield list(report.saved_reports.all()[0].data.rows)
-
-    if report.exists:
-        report.delete()
-
-
-@pytest.fixture(scope="module")
-def chargeback_report_custom(appliance, vm_ownership, assign_custom_rate, provider):
-    # Create a Chargeback report based on a custom rate; Queue the report
-    owner = vm_ownership
-    data = {
-        'menu_name': 'cb_custom_' + provider.name,
-        'title': 'cb_custom' + provider.name,
-        'base_report_on': 'Chargeback for Vms',
-        'report_fields': ['Memory Used', 'Memory Used Cost', 'Owner',
-        'CPU Used', 'CPU Used Cost',
-        'Disk I/O Used', 'Disk I/O Used Cost',
-        'Network I/O Used', 'Network I/O Used Cost',
-        'Storage Used', 'Storage Used Cost'],
-        'filter': {
-            'filter_show_costs': 'Owner',
-            'filter_owner': owner,
-            'interval_end': 'Today (partial)'
-        }
-    }
-    report = appliance.collections.reports.create(is_candu=True, **data)
-
-    logger.info('Queuing chargeback report with custom rate for {} provider'.format(provider.name))
-    report.queue(wait_for_finish=True)
-
-    yield list(report.saved_reports.all()[0].data.rows)
-
-    if report.exists:
-        report.delete()
-
-
-@pytest.fixture(scope="module")
-def new_compute_rate(appliance):
-    # Create a new Compute Chargeback rate
-    try:
-        desc = 'cstm_' + fauxfactory.gen_alphanumeric()
-        compute = appliance.collections.compute_rates.create(
-            description=desc,
-            fields={
-                'Used CPU': {'per_time': 'Hourly', 'variable_rate': '3'},
-                'Used Disk I/O': {'per_time': 'Hourly', 'variable_rate': '2'},
-                'Used Memory': {'per_time': 'Hourly', 'variable_rate': '2'}
-            }
-        )
-        storage = appliance.collections.storage_rates.create(
-            description=desc,
-            fields={
-                'Used Disk Storage': {'per_time': 'Hourly', 'variable_rate': '3'}
-            }
-        )
-    except Exception as ex:
-        pytest.fail(
-            'Exception while creating compute/storage rates for chargeback report tests. {}'
-            .format(ex)
-        )
-
-    yield desc
-
-    for entity in [compute, storage]:
-        try:
-            entity.delete_if_exists()
-        except Exception as ex:
-            pytest.fail(
-                'Exception while cleaning up compute/storage rates for chargeback report tests. {}'
-                .format(ex)
-            )
+COST_DEVIATION = 1
+RESOURCE_DEVIATION = 0.25
+
+
+def compare_values(estimated_value, report_value, is_cost=True):
+    if is_cost:
+        deviation = COST_DEVIATION
+        report_value = report_value.replace('$', '').replace(',', '')
+    else:
+        deviation = RESOURCE_DEVIATION
+        report_value = report_value.replace('GB', '').replace('MB', '').replace(' ', '')
+    return (estimated_value - deviation <= float(report_value)
```

Comment:
Hitting a TypeError here:
```>       assert compare_values(estimated_value, report_value), ("Reported network usage cost "
            f" {report_value} does not match expected value {estimated_value}.")

cfme/tests/intelligence/reports/test_validate_chargeback_report.py:144: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

estimated_value = None, report_value = '1.00', is_cost = True

    def compare_values(estimated_value, report_value, is_cost=True):
        if is_cost:
            deviation = COST_DEVIATION
            report_value = report_value.replace('$', '').replace(',', '')
        else:
            deviation = RESOURCE_DEVIATION
            report_value = report_value.replace('GB', '').replace('MB', '').replace(' ', '')
>       return (estimated_value - deviation <= float(report_value)
            <= estimated_value + deviation)
E       TypeError: unsupported operand type(s) for -: 'NoneType' and 'int'

cfme/tests/intelligence/reports/test_validate_chargeback_report.py:68: TypeError```

Reply:
Fixed. I was accidentally re-using the same dict object for each 'rate_tier' in the list 'rates' created in the resource_cost fixture. Moved "rate_tier = {}" into the loop over each row in the query output.